[ACM-11953]: Remove/replace references to hypershift CLI (#6521)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/manage_aws_infra_iam.adoc

[#hosting-cluster-create-aws-infra]
=== Creating the AWS infrastructure

> To create the AWS infrastructure, use an infrastructure tool such as Terraform, Pulumi, or your own automation to create a YAML file that includes the following fields:

Reviewer: ```suggestion
To create the AWS infrastructure, use your favorite infrastructure automation/provisioning tool to create a YAML file that includes the following fields:
```

Reviewer: Also we need to describe what exactly needs creating.

Author: I could use some help with that next week.

Author: @csrwng @enxebre While Adel is out, can either of you help me provide more details here? Based on his comment above, it sounds like I need to describe exactly what the user has to create when they're creating the AWS infrastructure.

[#hosting-cluster-create-aws-infra]
=== Creating the AWS infrastructure

> To create the AWS infrastructure, use your favorite infrastructure automation and provisioning tool to create a YAML file. Ensure that the file includes the following fields:

Reviewer 2: s/to create a YAML file. Ensure that the file includes the following fields:/to create a VPC for your cluster. You can also use the AWS console. Ensure that the VPC includes private and public subnets and resources for external access such as NAT gateway and internet gateway./

I would also add:
In addition to the VPC, you need a private Hosted Zone for your cluster's ingress. If creating clusters that use PrivateLink (Private or PublicAndPrivate access modes), you need an additional hosted zone for PrivateLink.

And remove the list of resources below.

Author: Done

[#hosting-cluster-create-aws-infra]
=== Creating the AWS infrastructure

> To create the AWS infrastructure, use your favorite infrastructure automation and provisioning tool to create a VPC for your cluster. You can also use the AWS console. 

Reviewer 2: I would add a link for the AWS console instructions: https://docs.aws.amazon.com/vpc/latest/userguide/create-vpc.html#create-vpc-and-other-resources

Author: ~Will do~
Done

> To create the AWS infrastructure, use your favorite infrastructure automation and provisioning tool to create a VPC for your cluster. You can also use the AWS console. 

Reviewer 3: >To create the AWS infrastructure, use your favorite infrastructure automation and provisioning tool to create a VPC for your cluster. You can also use the AWS console.

are we specifying anywhere which infrastructure is needed to be created with the tool?

Author: ~@enxebre Not that I know of. If it needs to be a specific type of infrastructure, let me know and I can add that to the docs.~

Author: I updated this section to state that you need to create a VPC and other resources, and pointed to the AWS docs.

  name: <node_pool_name> <10>
  namespace: clusters
spec:
  arch: amd64
  clusterName: <hosted_cluster_name>
  management:
    autoRepair: true
    upgradeType: Replace
  nodeDrainTimeout: 0s
  platform:
    aws:
      instanceProfile: <instance_profile_name> <11>
      instanceType: m6i.xlarge
      rootVolume:
        size: 120
        type: gp3
      subnet:
        id: <subnet_xxx>
    type: AWS
  release:
    image: quay.io/openshift-release-dev/ocp-release:4.16-x86_64
  replicas: 2
status:
  replicas: 0
----

<1> Replace `<pull_secret_name>` with the name of your pull secret. 
<2> Replace `<etcd_encryption_key_name>` with the name of your etcd encryption key.
<3> Replace `<ssh_key_name>` with the name of your SSH key.
<4> Replace `<hosted_cluster_name>` with the name of your hosted cluster.
<5> Replace `<dns_domain>` with your base DNS domain, such as `example.com`.
<6> Replace `<infra_id>` with the value that identifies the IAM resources that are associated with the hosted cluster.
<7> Replace `<issuer_url>` with your issuer URL, which ends with your `infra_id` value. For example, `https://example-hosted-us-west-1.s3.us-west-1.amazonaws.com/example-hosted-infra-id`.

> <8> Replace `<subnet_xxx>` with your subnet ID. Both private and public subnets need to be tagged. For public subnets, use `kubernetes.io/role/elb=1`. For private subnets, use `kubernetes.io/role/internal-elb=1`.

Author: @csrwng I updated line 649 since your last review. Let me know if anything should change.

Reviewer 2: Looks good, thx!

<9> Replace `<vpc_xxx>` with your VPC ID.
<10> Replace `<node_pool_name>` with the name of your `NodePool` resource.
<11> Replace `<instance_profile_name>` with the name of your AWS instance.

[#hosting-cluster-create-aws-iam]
=== Creating the AWS IAM resources

In AWS, you need to create the following IAM resources:

* One OIDC provider, which is required to enable STS authentication
* Seven roles, which are separate for every component that interacts with the provider, such as the Kubernetes controller manager, cluster API provider, and registry
* One instance profile, which is the profile that is assigned to all worker instances of the cluster

. To create an OIDC provider, follow the steps in link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html[Create an OpenID Connect (OIDC) identity provider in IAM] in the AWS documentation.

> . To create the roles for each component that interacts with the provider, follow the procedures in link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create.html[Creating IAM roles] in the AWS documentation.

Reviewer 3: how would a user know which roles need to be created and which policies do they need? don't we want to share examples here? like the ones generated by the dev cli?

Author: ~I think sharing examples here would be great, but I need help to know what to share. If you can point me in the right direction, that would be very helpful.~

Author: A section earlier in this file goes into detail about the roles. I added a reference to that section on line 666.

* Two route tables: 1 private and 1 public
* Two private hosted zones: 1 for cluster Ingress and 1 for PrivateLink, in case you create a private cluster

All of those resources contain the `kubernetes.io/cluster/INFRA_ID=owned` tag, where `INFRA_ID` is the value that you specified in the command.
All of those resources contain the `kubernetes.io/cluster/<infra_id>=owned` tag, where `<infra_id>` is the value that you specified in the command.

[#hosting-cluster-create-aws-iam]
=== Creating the AWS IAM resources

To create the AWS IAM resources, enter the following command:

> //lahinson - july 2023 - update hypershift cli command here

Author: The following example YAML file is based on an example that Juan  provided in a Slack thread: https://redhat-internal.slack.com/archives/G01QS0P2F6W/p1715339148437149

  name: <etcd_encryption_key_name> <2>
  namespace: clusters
type: Opaque
---
apiVersion: v1
data:
  id_rsa: xxxxxxxxx
  id_rsa.pub: xxxxxxxxx
kind: Secret
metadata:
  creationTimestamp: null
  labels:
    hypershift.openshift.io/safe-to-delete-with-cluster: "true"
  name: <ssh-key-name> <3>
  namespace: clusters
---
apiVersion: hypershift.openshift.io/v1beta1
kind: HostedCluster
metadata:
  creationTimestamp: null
  name: <hosted_cluster_name> <4>
  namespace: clusters
spec:
  autoscaling: {}
  configuration: {}
  controllerAvailabilityPolicy: SingleReplica
  dns:
    baseDomain: <dns_domain> <5>
    privateZoneID: xxxxxxxx
    publicZoneID: xxxxxxxx
  etcd:
    managed:
      storage:
        persistentVolume:
          size: 8Gi
          storageClassName: gp3-csi
        type: PersistentVolume
    managementType: Managed
  fips: false
  infraID: <infra_id> <6>
  issuerURL: <issuer_url> <7>
  networking:
    clusterNetwork:
    - cidr: 10.132.0.0/14
    machineNetwork:
    - cidr: 10.0.0.0/16
    networkType: OVNKubernetes
    serviceNetwork:
    - cidr: 172.31.0.0/16
  olmCatalogPlacement: management
  platform:
    aws:
      cloudProviderConfig:
        subnet:

>           id: subnet-04839d7ee947c9c5a

Reviewer 4: I believe this would be a field that needs replaced, 'subnet-xxx'

>         vpc: vpc-008885fd2a9502c66

Reviewer 4: I believe this would be a filed that needs replaced, 'vpc-xxx'

        zone: us-west-1b
      endpointAccess: Public

>       multiArch: false

Reviewer 4: Should we switch this example to show this as true? All HCs in ROSA going forward will be multiArch enabled.

Author: Not sure. @csrwng or @enxebre can you answer Bryan's question?

Reviewer 2: Right now the default of the hcp cli is to set it to false, so I wouldn't set it to true here (until we change the default).

Reviewer 4: We are planning to set it to true by default in https://github.com/openshift/hypershift/pull/4272

Author: @bryan-cox Just wanted to check on this one -- should I change line 577 to `true`?

Reviewer 4: 569? This is still correct `multiArch: false`

      region: us-west-1
      rolesRef:
        controlPlaneOperatorARN: arn:aws:iam::820196288204:role/<infra_id>-control-plane-operator
        imageRegistryARN: arn:aws:iam::820196288204:role/<infra_id>-openshift-image-registry
        ingressARN: arn:aws:iam::820196288204:role/<infra_id>-openshift-ingress
        kubeCloudControllerARN: arn:aws:iam::820196288204:role/<infra_id>-cloud-controller
        networkARN: arn:aws:iam::820196288204:role/<infra_id>-cloud-network-config-controller
        nodePoolManagementARN: arn:aws:iam::820196288204:role/<infra_id>-node-pool
        storageARN: arn:aws:iam::820196288204:role/<infra_id>-aws-ebs-csi-driver-controller
    type: AWS
  pullSecret:
    name: <pull_secret_name>
  release:

>     image: quay.io/openshift-release-dev/ocp-release:4.15.9-x86_64

Reviewer 4: Should we use a 4.16 image here now that its GA?

Author: Yes. Good catch! Will fix.

Author: @bryan-cox I replaced it with `ocp-release:4.16-x86_64`. Let me know if that's not quite right.

  secretEncryption:
    aescbc:
      activeKey:
        name: <etcd_encryption_key_name>
    type: aescbc
  services:
  - service: APIServer
    servicePublishingStrategy:
      type: LoadBalancer
  - service: OAuthServer
    servicePublishingStrategy:
      type: Route
  - service: Konnectivity
    servicePublishingStrategy:
      type: Route
  - service: Ignition
    servicePublishingStrategy:
      type: Route
  - service: OVNSbDb
    servicePublishingStrategy:
      type: Route
  sshKey:
    name: <ssh_key_name>
status:
  controlPlaneEndpoint:
    host: ""
    port: 0
---
apiVersion: hypershift.openshift.io/v1beta1
kind: NodePool
metadata:
  creationTimestamp: null
  name: <node_pool_name> <8>
  namespace: clusters
spec:
  arch: amd64
  clusterName: <hosted_cluster_name>
  management:
    autoRepair: true
    upgradeType: Replace
  nodeDrainTimeout: 0s
  platform:
    aws:
      instanceProfile: <instance_profile_name> <9>
      instanceType: m6i.xlarge
      rootVolume:
        size: 120
        type: gp3
      subnet:
        id: subnet-04839d7ee947c9c5a
    type: AWS
  release:

>     image: quay.io/openshift-release-dev/ocp-release:4.15.9-x86_64

Reviewer 4: Should we use a 4.16 image here now that its GA?

Author: I replaced it with `ocp-release:4.16-x86_64`. Let me know if that should change.

  replicas: 2
status:

>   replicas: 0

Reviewer: It needs to be reviewed by engineering and templatizing what needs to be templatized.

Author: Yes, please. I assigned Bryan as a reviewer. If I should include anyone else (maybe Juan?), let me know.

Reviewer: I'd recommend Cesar or Alberto since this is the AWS provider.

<9> Replace `<instance_profile_name>` with the name of your AWS instance.

[#hosting-cluster-create-aws-iam]
=== Creating the AWS IAM resources

In AWS, you need to create the following IAM resources:

* One OIDC provider, which is required to enable STS authentication
* Seven roles, which are separate for every component that interacts with the provider, such as the Kubernetes controller manager, cluster API provider, and registry
* One instance profile, which is the profile that is assigned to all worker instances of the cluster

. To create an OIDC provider, follow the steps in link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html[Create an OpenID Connect (OIDC) identity provider in IAM] in the AWS documentation.

. To create the roles for each component that interacts with the provider, follow the procedures in link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create.html[Creating IAM roles] in the AWS documentation.

. To create an instance profile, see link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html[Using instance profiles] in the AWS documentation.

> 

Reviewer: Review needed here by engineering.

Reviewer: What other infrastructure components are needed? What if it's a private cluster?

Author: Some background here, in case it helps.

The original text mentioned creating the following resources:
- an OIDC provider
- A role for each component that interacts with the provider
- An instance profile

In the original text, commands were provided to create each resource. I replaced the commands with the equivalent instructions in the AWS docs. If we need to mention other infrastructure components, I'm happy to do that, but I'd need guidance on what to add.

Reviewer: I understand, but since we are removing the abstraction, we need to be sure we document all dependencies otherwise it won't be as intuitive to create a cluster.

Author: @csrwng Do we need to mention any other infrastructure components here?

Reviewer 2: No, since this is the section on IAM and not infrastructure. The infrastructure section where we tell you to create a VPC should be sufficient.

* Two route tables: 1 private and 1 public
* Two private hosted zones: 1 for cluster Ingress and 1 for PrivateLink, in case you create a private cluster

> All of those resources contain the `kubernetes.io/cluster/<infra_id>=owned` tag, where `<infra_id>` is the value that you specified in the command.

Reviewer 2: I would remove the above, since it's not required.
However, we do need the public and private subnets to be tagged:
(from https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html)
Public subnet: `kubernetes.io/role/elb=1`
Private subnet: `kubernetes.io/role/internal-elb=1`

All of those resources contain the `kubernetes.io/cluster/INFRA_ID=owned` tag, where `INFRA_ID` is the value that you specified in the command.

.Example YAML file
[source,yaml]
[subs="+quotes"]

> ----

Author: Need to confirm whether this whole snippet is necessary to include. It's quite long -- we want to include necessary content, but don't want to hit people with a wall of text.

    aescbc:
      activeKey:
        name: <etcd_encryption_key_name>
    type: aescbc
  services:
  - service: APIServer
    servicePublishingStrategy:
      type: LoadBalancer
  - service: OAuthServer
    servicePublishingStrategy:
      type: Route
  - service: Konnectivity
    servicePublishingStrategy:
      type: Route
  - service: Ignition
    servicePublishingStrategy:
      type: Route
  - service: OVNSbDb
    servicePublishingStrategy:
      type: Route
  sshKey:
    name: <ssh_key_name>
status:
  controlPlaneEndpoint:
    host: ""
    port: 0
---
apiVersion: hypershift.openshift.io/v1beta1
kind: NodePool
metadata:
  creationTimestamp: null
  name: <node_pool_name> <8>
  namespace: clusters
spec:
  arch: amd64
  clusterName: <hosted_cluster_name>
  management:
    autoRepair: true
    upgradeType: Replace
  nodeDrainTimeout: 0s
  platform:
    aws:
      instanceProfile: <instance_profile_name> <9>
      instanceType: m6i.xlarge
      rootVolume:
        size: 120
        type: gp3
      subnet:
        id: subnet-04839d7ee947c9c5a
    type: AWS
  release:
    image: quay.io/openshift-release-dev/ocp-release:4.15.9-x86_64
  replicas: 2
status:
  replicas: 0
----

> 

Author: We need call out text for the various user-replaced variables in the YAML.
